README.md
### 分布式秒级端口扫描系统

&emsp;&emsp;分布式秒级端口扫描系统，用于快速巡检公司服务器对外的暴露面，此为应对海量服务器!

&emsp;&emsp;部署方式参考我的博客：https://code3322.cn/2022/03/Celery/

&emsp;&emsp;项目使用python3 + celery + redis + mysql来实现，具体看部署文档。

ChangeLog.md
### ChangeLog

&emsp;
#### 修订记录
| 版本号 | 更新时间   | 更新描述  | 其它 |
| --: | --: | --: | --: | 
| v1.0 | 2022.3.15 | 初步开发完成 |  |
|    |    |    |    |

&emsp;
#### [v.10] 2022.3.15
初步开发完成

遗留问题:
* 读取ip列表只实现了读文本, 暂未实现mysql, cmdb接口
* 消息中间件只实现了redis, 暂未实现rabbitmq、kafka

dstb_scan_system/core_code/nmap_scan.py
#!/usr/bin/env python
#encoding: utf-8

#-------------------------------------------------------------------------
#头文件
from core_code.scan_core import sec_scan
import time
import nmap
import json
import pymysql
import datetime


#-------------------------------------------------------------------------
#全局变量
# mysql数据库连接信息
MysqlHost = '127.0.0.1'
MysqlUser = 'user'
MysqlPwd = 'pwd'
MysqlDBName = 'xxxx'


#-------------------------------------------------------------------------
#类、函数声明与定义
@sec_scan.task
def nmap_service_detect(ip=None, port_list=None, args="-sS -n -sV"):
    #显示一下传过来的参数,调试用
    #print('ip: %s, port list: %s, args: %s' % (ip, port_list, args))

    #执行Nmap并进行扫描
    nm = nmap.PortScanner()
    nm.scan(ip, port_list, args)
    
    #显示被扫描的ip列表,调试用
    #print(nm.all_hosts())

    #显示被扫描的端口列表,调试用
    #print(nm[ip].all_tcp())

    #端口列表排序
    lport = nm[ip].all_tcp()
    lport.sort()

    #分析端口的服务版本信息
    result_list = []
    result_list.append(ip)
    for line in lport:
        if nm[ip].tcp(line)['state'] == 'open':
            #print('ip: %s, port: %d, info: %s, vesion: %s %s' % (ip, line, nm[ip].tcp(line)['name'], nm[ip].tcp(line)['product'], nm[ip].tcp(line)['version']))
            info_dict = {}
            info_dict['port'] = line
            info_dict['info'] = str(nm[ip].tcp(line)['name'])
            info_dict['version'] = str(nm[ip].tcp(line)['product']) + str(nm[ip].tcp(line)['version'])
            result_list.append(info_dict)

    #分析结果, 新数据就入库, 这里可以不用执行, 因为消息中间件已经存储结果,mysql插入数据比较慢, 这里会拖慢进度, 比较懒就另当别论
    if len(result_list) != 0:
        detect_data_isnew_tmp = "select count(*) from scan_result where ip='%s' and port = %d"
        i = 1
        while i < len(result_list):
            detect_sql = detect_data_isnew_tmp % (ip, result_list[i]['port'])
            detect_data = sql_exec(detect_sql, return_data=True)
            if not detect_data[0][0]:
                insert_sql_tmp = "insert into scan_result(ip, port, servicename, version, tasktype, submitter, flags) value('%s', '%d', '%s', '%s', '%s', '%s', '%d')"
                insert_sql = insert_sql_tmp % (ip, result_list[i]['port'], result_list[i]['info'], result_list[i]['version'], 'person', 'xxx', 0)
                sql_exec(insert_sql, return_data=False)
            i += 1

        return result_list


#把结果插入数据库
def sql_exec(sql=None, return_data=False):
    '''执行mysql 语句'''
    conn = pymysql.connect(host=MysqlHost, user=MysqlUser, passwd=MysqlPwd, db=MysqlDBName, port=3306, charset="utf8", connect_timeout=10)
    cursor = conn.cursor()
    cursor.execute(sql)
    conn.commit()
    
    data = cursor.fetchall()
    if return_data:
        return data
    else:
        pass
    
    conn.close()


dstb_scan_system/core_code/scan_core.py
#!/usr/bin/env python
#encoding: utf-8

#-------------------------------------------------------------------------
#头文件
from celery import Celery


#-------------------------------------------------------------------------
#全局变量


#-------------------------------------------------------------------------
#类、函数声明与定义
#参数include是worker中要处理消息的函数, 需要导出, 导出已经放在scan_config.py中了, 此代码作为参考, core_code为你的项目的名, 也就是目录名, masscan为core_code目录中要执行的python文件名, 导出路径到文件名就行, 不需要连函数名也导出
#sec_scan = Celery("secScan", include = ["core_code.masscan_scan", "core_code.nmap_scan"])

#sec_scan这个是类的对象, 在任务函数中要使用
sec_scan = Celery("secScan") 

#scan_config就是这个这个目录下的scan_config.py文件
sec_scan.config_from_object("core_code.scan_config")


#-------------------------------------------------------------------------
#主函数
#如果不写以下两句, 需要在命令行使用: celery -A core_code.scan_core worker -B -l info 来执行
if __name__ == '__main__':
    sec_scan.start()


dstb_scan_system/core_code/scan_config.py
#!/usr/bin/env python
#encoding: utf-8

#-------------------------------------------------------------------------
#头文件
from datetime import timedelta
from celery.schedules import crontab


#-------------------------------------------------------------------------
#全局变量
#消息输入的队列, 依情况修改, 此使用redis 0这个库
BROKER_URL = "redis://127.0.0.1:6379/0"

#消息输出的队列, 依情况修改, 使用redis 1这个库
CELERY_RESULT_BACKEND = "redis://127.0.0.1:6379/1"

#时区设置
CELERY_TIMEZONE='Asia/Shanghai'

#导出任务, 手动调用、计划任务, 都要导出, 路径只指定到文件名, 里面的函数需要消息发送端来导出并使用
#CELERY_IMPORTS = ('core_code.masscan_scan', 'core_code.nmap_scan', 'core_code.cron_masscan_scan', 'core_code.corn_nmap_scan') 
CELERY_IMPORTS = ('core_code.masscan_scan', 'core_code.nmap_scan')


#定时任务调度器
CELERYBEAT_SCHEDULE={
    #任务名字, 可自定义, 无关联
    'masscan_cron':{
        #任务启动的函数, 指定要具体函数
        'task':'core_code.cron.cron_masscan_scan.mass_scan',
        #定时时间设置，此为每10秒一次
        'schedule':timedelta(seconds=5),
        #传递给函数的参数
        'args':(['127.0.0.1', '1-65535', '-n -pn'])
    },
}

'''
#定时任务调度器, 此为多个任务, 此为参考代码
CELERYBEAT_SCHEDULE={
    #任务名字, 可自定义, 无关联
    'task1':{
        # 任务启动的函数
        'task':'celery_app.task1.add',
        # 定时时间设置，每10秒一次
        'schedule':timedelta(seconds=10),
        # 传递的参数
        'args':(2,8)
    },
    'task2':{
        'task':'celery_app.task2.mul',
        # 定时时间设置，16:45
        'schedule':crontab(hour=16,minute=45),
        'args':(4,5)
    }
}
'''


dstb_scan_system/core_code/masscan_scan.py
#!/usr/bin/env python
#encoding: utf-8

#-------------------------------------------------------------------------
#头文件
from core_code.scan_core import sec_scan
import time
import masscan


#-------------------------------------------------------------------------
#全局变量


#-------------------------------------------------------------------------
#类、函数声明与定义
@sec_scan.task
def mass_port_scan(ip=None, port_list='1-65535', args='-n -pn --rate 1000'):
    mass = masscan.PortScanner()
    scan_result = mass.scan(ip, ports=port_list, arguments=args)

    result = []
    for key in scan_result['scan'][ip]['tcp'].keys():
        state = scan_result['scan'][ip]['tcp'].get(key)['state']
        if ('open' == state):
            result.append(key)

    if len(result):
        result.insert(0, ip)
    else:
        fail_result = ip + 'All ports is closed'
        return  fail_result

    return result
        
        


dstb_scan_system/port_scan_produce.py
#!/usr/bin/env python
#encoding: utf-8


#-------------------------------------------------------------------------
#头文件
import logging
import time
import sys
import datetime
from core_code.masscan_scan import mass_port_scan
from celery.result import AsyncResult
from core_code.nmap_scan import nmap_service_detect


#-------------------------------------------------------------------------
#全局变量
#扫描ip来源, 可为mysql、file、cmdb
#mysql<暂时没实现，后续实现>
MysqlHost = '127.0.0.1'
MysqlUser = 'xxxx'
MysqlPwd = 'xxxx'
MysqlDB = 'xxxx'

#file
ipListName = 'ip.txt'

#运行时日志文件名
logFile = 'error.log'

#运行时日志级别
log_level = 'logging.DEBUG'

#默认要扫描的端口列表,精选经常被用到的业务组件对应的端口,也可以改ip.txt里面的第二个字段
scan_ports = '21,22,23,992,25,53,88,110,111,80,443,135,137,138,139,161,389,636,445,512,873,1025,1433,1723,1512,2049,2181,2375,3128,3306,3389,3690,4899,5432,5631,5900,6379,7001,7002,8000,8012,8069,8080,8649,9043,9200,9300,10250,11211,27017,50000,50070,50075,50090'

#mass默认参数,没事也别改,因为过滤数据时可能会出错,也可改ip.txt里面的第三个字段
mass_arg = '-n -Pn --rate 1000'

#任务成功计数,有需要可导出到看板
success_num = 0

#任务失败计数,有需要可导出到看板
fail_num = 0

#本次要扫描的总ip数
count_num = 0

#是否需要调用nmap来认别服务,mass扫描结果保存的消息中间件,nmap的结果保存在mysql
Nmap_Service_Scan_Trun = True


#-------------------------------------------------------------------------
#类、函数声明与定义
def exec_mass_task(src=None):
    '''执行任务,src表来ip的来源, 可为mysql、cmdb、文本(file)'''
    if src == 'file':
        exec_src_file() 
    elif src == 'mysql':
        '''后续实现'''
        pass
    elif src == 'cmdb':
        '''后续实现'''
        pass
    else:
        sys.exit(0)


def exec_src_file(portlist=None, arg=None):
    '''从文本读取ip并执行,扫描的端口与参数即可用全局变量定义的,也可依实际情况改ip.txt中的格式'''
    task_list = []
    with open(ipListName) as FileOjb:
        for line in FileOjb:
            global count_num
            count_num += 1
            #处理扫描参数
            final_ip = line.split("#")[0].strip()
            final_ports = line.split("#")[1].strip()
            if len(final_ports) == 0:
                final_ports = scan_ports
            final_args = line.split("#")[2].strip()
            if len(final_args) == 0:
                final_args = mass_arg

            #print('ip: %s, ports: %s, args: %s' % (final_ip, final_ports, final_args))
            result = mass_port_scan.delay(final_ip, final_ports, final_args)

            #把执行id存储起来, 用于状态的判断
            ip_dict = {}
            ip_dict['ip'] = final_ip
            ip_dict['task_id'] = str(result.id)
            ip_dict['status'] = str(result.status)
            task_list.append(ip_dict)

    #判断任务的状态,如果成功,调用nmap执行服务认识, 如果失败进行记录
    while True:
        time.sleep(5) 
        #获取任务的结果
        for line in task_list:
            get_ret = AsyncResult(line['task_id'])
            if get_ret.successful():
                global success_num
                success_num += 1
                tmp = get_ret.get()

                #传给Nmap函数扫描的函数
                nmap_scan_ip = str(tmp[0]).strip()
                nmap_scan_ports = str(tmp[1:]).strip()

                #masscan任务执行成功, 接着可调用nmap来认识服务
                if Nmap_Service_Scan_Trun:
                    nmap_result = nmap_service_detect.delay(nmap_scan_ip, nmap_scan_ports)

                #任务执行成功,结果已经获取,不需要再循环查看,从获取队列中删除此任务
                task_list = [item for item in task_list if not item['task_id'] == line['task_id']]
            elif get_ret.failed():
                global fail_num 
                fail_num += 1
                #任务执行失败,结果已经获取,不需要再循环查看,从获取队列中删除此任务
                task_list = [item for item in task_list if not item['task_id'] == line['task_id']]

        if len(task_list) == 0:
            break

    print('The Tasks have: %d' % (count_num))
    print('Success scan: %d' % (success_num))
    print('Fail scan: %d' % (fail_num))



def exec_src_mysql():
    '''从mysql中读取ip并执行, 后续实现'''
    pass

def exec_src_cmdb():
    '''从cmdb读取ip并执行, 后续实现'''
    pass


#-------------------------------------------------------------------------
#主函数
if __name__ == '__main__':
    '''程序运行时的msg'''
    starttime = datetime.datetime.now()
    print('-------------------------------------------------------------------------')
    print('程序将很快执行，现在时间是:',starttime.strftime('%Y-%m-%d %H:%M:%S'))
    print('如遇异常，请查看当前目录下的日志文件 error.log')
    print('正在进行程序初始化, 请稍等.....')
    print('')


    #运行时日志设置
    logger = logging.getLogger()
    logger.setLevel(logging.DEBUG)
    handle = logging.FileHandler(logFile, mode='a+')
    handle.setLevel(logging.DEBUG)    
    formatter = logging.Formatter('%(asctime)s - %(filename)s - [line:%(lineno)d]  -  %(levelname)s - %(message)s')
    handle.setFormatter(formatter)
    logger.addHandler(handle)


    #核心代码执行
    try:
        #测试任务代码
        #mass_port_scan.delay('61.152.109.30', '22,80,443,3306', '-n -Pn')
        
        #执行代码
        exec_mass_task(src='file') 

    except Exception as e:
        logger.error(str(e), exc_info=True)
        sys.exit(0)


    '''程序运行结束，并显示运行了多久'''
    endtime = datetime.datetime.now()
    seconds = (endtime - starttime).seconds
    print('程序执行完成，一共花费了 %d 秒' % (seconds))
    print('-------------------------------------------------------------------------')
